File: 142_LLcycle2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        if(head == nullptr)
            return nullptr;
        
        if(head->next == nullptr)
            return nullptr;
        
        ListNode *slow = head,
                 *fast = head;
     
        while(fast != nullptr && fast->next != nullptr){
            slow = slow->next;
            fast = fast->next->next;
            
            if(slow == fast)
                return findCycleStart(head, fast);
        }        
        return nullptr;
    }
    
    ListNode *findCycleStart(ListNode *head, ListNode *meetNode){
        // One ptr start from head, one ptr start from the encountering node
        // Both move forward by one, then they'll meet at the start of cycle
        ListNode* walk = head;
        
        while(walk != meetNode){
            walk = walk->next;
            meetNode = meetNode->next;
        }
        
        return walk;
    }
};

File: 142_LL-cycle-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head
        while fast and fast.next:          
            slow = slow.next
            if (fast := fast.next.next) == slow: 
                temp = head
                while temp != slow:
                    temp = temp.next
                    slow = slow.next
                return temp    
        return None
